Use type equality, not base_type_eq in local safe pointers

tautschnig opened the pull request:
We no longer need to resort to tag/symbol type resolution. This enables partial
revert of 959c7a5f7f653a7 (Bugfix: Maintain safe_pointers per-path) as we no
longer need to maintain a namespacet in local_safe_pointerst. On SV-COMP's
ReachSafety-ECA, copying safe_pointers accounted for 14% of the time spent in
goto_symext::symex_goto (715 of 5119 seconds).

~Requires #3966 to be merged for Java tests to pass.~

<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.
--->

- [x] Each commit message has a non-empty body, explaining why the change was made.
- n/a Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- n/a The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [x] My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- n/a White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

smowton reviewed (approved):
When tests pass, of course. Thanks for removing the pointless copying of `local_safe_pointers`, I'd been meaning to fix that!

thk123 commented on src/goto-symex/symex_function_call.cpp:
```diff
@@ -240,7 +240,7 @@ void goto_symext::symex_function_call_code(
 
   auto emplace_safe_pointers_result =
-    state.safe_pointers.emplace(identifier, local_safe_pointerst{ns});
+    path_storage.safe_pointers.emplace(identifier, local_safe_pointerst{});
```
:pick: AFAICT this variable renaming should be done in a seperate commit, ideally with an explanation of why this name is more accurate (I don't know this code well enough to comment on whether this is a good renaming or not). :dove:

tautschnig replied:
Ah, no, this isn't variable renaming, it's an ownership change. `safe_pointers` used to belong to `goto_symex_statet`, but is now moved to `path_storaget`.

thk123 commented on src/goto-symex/symex_main.cpp:
```diff
@@ -178,7 +178,7 @@ void goto_symext::initialize_entry_point(
 
   auto emplace_safe_pointers_result =
-    state.safe_pointers.emplace(function_id, local_safe_pointerst{ns});
+    path_storage.safe_pointers.emplace(function_id, local_safe_pointerst{});
```
:pick: :dove:

thk123 commented on src/goto-symex/path_storage.h:
```diff
@@ -90,6 +92,8 @@ class path_storaget
   /// Counter for nondet objects, which require unique names
   symex_nondet_generatort build_symex_nondet;
 
+  std::unordered_map<irep_idt, local_safe_pointerst> safe_pointers;
```
:pick: You might choose to doxygen this variable, `safe_pointers` doesn't mean mich to me, but perhaps just a reference to a definition of a "safe pointer" would be a good addition here.

tautschnig replied:
Documentation added.

thk123 commented on src/goto-symex/symex_dereference.cpp:
```diff
@@ -262,9 +262,8 @@ void goto_symext::dereference_rec(
-        expr_is_not_null =
-          state.safe_pointers.at(expr_function).is_safe_dereference(
-            to_check, state.source.pc);
+        expr_is_not_null = path_storage.safe_pointers.at(expr_function)
```
:dove:

thk123 commented on src/analyses/local_safe_pointers.h:
```diff
@@ -23,45 +23,17 @@ Author: Diffblue Ltd
-  /// their subexpressions, are identical except that one may use a symbol_typet
-  /// while the other uses that type's expanded (namespacet::follow'd) form.
-  class base_type_comparet
+  class type_comparet
```
:question: Does this class even need to exist? Except the ordering is on the exprs, unless the types match - this is somewhat surprising. Out of scope of this PR, but I wonder if this is really what's needed here...

tautschnig replied:
It now still is a class, but with a near-trivial inline implementation of the comparison. What approach other than a class would you have thought of?

thk123 commented on src/goto-symex/goto_symex_state.h:
```diff
@@ -317,7 +315,6 @@ inline goto_statet::goto_statet(const class goto_symex_statet &s)
     source(s.source),
     propagation(s.propagation),
     atomic_section_id(s.atomic_section_id),
-    safe_pointers(s.safe_pointers),
```
:pick: Ideally this removal would have gone in a different commit with an explanation of why not required.

tautschnig replied:
This PR is now composed of three commits:
1) Removing the `base_type_eq` comparison from `local_safe_pointerst`.
2) Small bits of cleanup in `local_safe_pointerst`.
3) Moving the storage from `goto_symex_statet` to `path_storaget`.

thk123 reviewed (approved):
Don't see anything blocking in this PR - though seems like quite a lot tangled up into one commit which made this a little difficult to review.

tautschnig committed:
Use type equality, not base_type_eq in local safe pointers

We no longer need to resort to tag/symbol type resolution.

tautschnig committed:
goto-symex: move safe_pointers to path storage

As local_safe_pointerst no longer needs a namespace we can do a partial revert
of 959c7a5f7f653a7 (Bugfix: Maintain safe_pointers per-path). On SV-COMP's
ReachSafety-ECA, copying safe_pointers accounted for 14% of the time spent in
goto_symext::symex_goto (715 of 5119 seconds).

tautschnig committed:
Code cleanup in local_safe_pointers

1) Inline (now trivial) comparision function.
2) Use skip_typecast.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 393da450e3d3f26082838071c11981a39a3d5839).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/100642937

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: e1bfc15191780d4198aaf1b8e77268f11eb46f7a).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/100810800
